### PR TITLE
domain: Unify lease

### PR DESCRIPTION
**Pull request by zimulala**

### CODE REVIEW

**Comment by zimulala** [comment]
PTAL @siddontang @qiuyesuifeng @shenli

File: domain/domain.go [hunk 1]
```diff
@@ -259,11 +256,19 @@ func NewDomain(store kv.Storage, lease time.Duration) (d *Domain, err error) {
...
 	d.infoHandle = infoschema.NewHandle(d.store)
+
 	d.ddl = ddl.NewDDL(d.store, d.infoHandle, &ddlCallback{do: d}, lease)
 	d.mustReload()
 
 	variable.RegisterStatistics(d)
 
+	// Only when the store is local that the lease value is 0.
+	// Set reload lease != 0 to ensure that loadSchemaInLoop can work normally.
+	// And if the store is local, reload lease is larger than ddl lease is ok.
```
Comment (shenli) [inline domain/domain.go]: // Local store allows reload lease greater than ddl lease.
// For other stores, reload lease is always equal to ddl lease. ?

File: domain/domain.go [hunk 2]
```diff
@@ -109,6 +109,11 @@ func (do *Domain) Store() kv.Storage {
 
 // SetLease will reset the lease time for online DDL change.
 func (do *Domain) SetLease(lease time.Duration) {
+	if lease <= 0 {
+		log.Infof("[ddl] SetLease %v failed, so do nothing", lease)
```
Comment (shenli) [inline domain/domain.go]: Is this an error? or a warning?

File: domain/domain.go (hunk 1 above)
Comment (zimulala) [inline domain/domain.go]: yep.

**Comment by zimulala** [comment]
PTAL @siddontang

**Comment by zimulala** [comment]
PTAL @shenli

**Comment by shenli** [comment]
LGTM

**Comment by shenli** [comment]
@qiuyesuifeng PTAL

**Comment by qiuyesuifeng** [comment]
LGTM. But perhaps for local store, it is needless to reload schema.

**Comment by zimulala** [comment]
PTAL @shenli @qiuyesuifeng @coocood

File: domain/domain.go [hunk 3]
```diff
@@ -115,13 +115,33 @@ func (do *Domain) Store() kv.Storage {
...
 		log.Warnf("[ddl] SetLease %v failed, so do nothing", lease)
 		return
 	}
 
-	do.leaseCh <- lease
+	do.m.Lock()
+	defer do.m.Unlock()
+	// It uses for test.
+	if lease == 0 {
+		if do.leaseCh != nil {
```
Comment (shenli) [inline domain/domain.go]: Add a warning here.

**Comment by shenli** [comment]
LGTM
@qiuyesuifeng PTAL

File: domain/domain.go [hunk 4]
```diff
@@ -115,8 +115,34 @@ func (do *Domain) Store() kv.Storage {
...
+		return
+	}
+
+	do.m.Lock()
+	defer do.m.Unlock()
+	// It uses for test.
+	if lease == 0 {
+		log.Warnf("[ddl] SetLease: %v", lease)
+		if do.leaseCh != nil {
+			close(do.leaseCh)
```
Comment (qiuyesuifeng) [inline domain/domain.go]: Why need close do.leaseCh here?

File: domain/domain.go [hunk 5]
```diff
@@ -224,9 +247,10 @@ func (do *Domain) loadSchemaInLoop(lease time.Duration) {
 			} else if err != nil {
 				log.Fatalf("[ddl] reload schema err %v", errors.ErrorStack(err))
 			}
-		case newLease := <-do.leaseCh:
-			if newLease <= 0 {
-				newLease = defaultLoadTime
+		case newLease, ok := <-do.leaseCh:
```
Comment (qiuyesuifeng) [inline domain/domain.go]: Why not using the origin code to handle set new lease but set do.leaseCh = nil.

File: domain/domain.go (hunk 4 above)
Comment (zimulala) [inline domain/domain.go]: Notify `loadSchemaInLoop` to exit the current goroutine。

**Comment by qiuyesuifeng** [comment]
LGTM

File: domain/domain.go [hunk 6]
```diff
@@ -115,8 +114,22 @@ func (do *Domain) Store() kv.Storage {
...
 func (do *Domain) SetLease(lease time.Duration) {
-	do.leaseCh <- lease
+	if lease <= 0 {
+		log.Warnf("[ddl] SetLease %v failed, so do nothing", lease)
+		return
+	}
+
+	do.m.Lock()
+	defer do.m.Unlock()
+	if do.leaseCh == nil {
```
Comment (qiuyesuifeng) [inline domain/domain.go]: Seems this logic can also be removed?

**Comment by zimulala** [comment]
PTAL @qiuyesuifeng

**Comment by qiuyesuifeng** [comment]
LGTM

**Pull request merged**